app/models/user.rb

```diff
@@ -17,19 +15,17 @@ class User < ActiveRecord::Base
          :confirmable, :lockable, :timeoutable,
          :omniauthable, :omniauth_providers => [:shibboleth]
 
-  def self.from_omniauth(auth)
-    user = User.find_by_email(auth.uid) || User.find_by(ccid: auth.uid)
-    if user and user.confirmed?
-      user.provider = auth.provider
-      user.ccid = auth.uid if (user.ccid != auth.uid)
-      return user
-    end
-    where(provider: auth.provider, email: auth.uid).first_or_create do |user|
-      user.skip_confirmation!
-      user.email = auth.uid
-      user.password = Devise.friendly_token[0,20]
-    end
-
+  scope :from_omniauth, ->(auth){ where('(email = ? OR ccid = ?) AND provider = ?', 
+                                         auth.uid, auth.uid, auth.provider).limit(1) }
```

Comment:
[Rails scopes](http://guides.rubyonrails.org/active_record_querying.html#scopes) are the preferred way of centralizing query logic in Rails models. They have a lot of nice properties, like composability and extensibility with all other user-written and built in Rails scopes, and many gems, that you otherwise have to do extra work to get if you use class methods to implement query functionality. So it's generally preferably to use scopes for containing query logic whenever possible.

->() is the Ruby 1.9 "stabby lambda" lambda syntax. In this case, we're declaring the scope to be implemented by a lambda that takes one argument, auth.